sdk/daml-script/test/daml/upgrades/ContractKeys.daml

```diff
@@ -98,6 +106,111 @@ exerciseUpdateKeyUnchanged = test $ do
   res === "V2"
 
+{- MODULE
+package: contract-key-upgrades
+contents: |
+  module ChangedKeyExpr where
+
+  data ChangedKeyExprKey = ChangedKeyExprKey with
+      p : Party
+      b : Bool
+    deriving (Eq, Show)
+
+  template ChangedKeyExpr
+    with
+      party : Party
+      b : Bool
+    where
+      signatory party
+      key (ChangedKeyExprKey party False) : ChangedKeyExprKey -- @V 1
+      key (ChangedKeyExprKey party b)     : ChangedKeyExprKey -- @V  2
+      maintainer key.p
+
+      choice ChangedKeyExprCall : Text
+        controller party
+        do pure "V1"                                          -- @V 1
+        do pure "V2"                                          -- @V  2
+
+  template ChangedKeyExprHelper
+    with
+      party : Party
+    where
+      signatory party
+
+      choice ChangedKeyExprFetch : ChangedKeyExpr with
+          cid : ContractId ChangedKeyExpr
+        controller party
+        do fetch cid
+
+      choice ChangedKeyExprFetchByKey : (ContractId ChangedKeyExpr, ChangedKeyExpr) with
+          k : ChangedKeyExprKey
+        controller party
+        do fetchByKey k
+
+      choice ChangedKeyExprExercise : Text with
+          k : ChangedKeyExprKey
+        controller party
+        do exerciseByKey @ChangedKeyExpr k ChangedKeyExprCall
+-}
+
+queryKeyChangedExprSameValue : Test
+queryKeyChangedExprSameValue = test $ do
+  a <- allocateParty "alice"
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a False)
+  keyRes <- queryContractKey a $ V2.ChangedKeyExprKey a False
+  case keyRes of
+    Some (foundCid, foundContract) | show foundCid == show cid && foundContract == V2.ChangedKeyExpr a False -> pure ()
+    _ -> assertFail $ "Didn't find correct contract, expected " <> show (cid, V2.ChangedKeyExpr a False) <> ", got " <> show keyRes
+
+queryKeyChangedExprChangedValue : Test
+queryKeyChangedExprChangedValue = test $ do
+  a <- allocateParty "alice"
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a True)
+  -- the following query works, even though the key value changed!
+  keyRes <- queryContractKey a $ V2.ChangedKeyExprKey a False
+  case keyRes of
+    Some (foundCid, foundContract) | show foundCid == show cid && foundContract == V2.ChangedKeyExpr a True -> pure ()
+    _ -> assertFail $ "Didn't find correct contract, expected " <> show (cid, V2.ChangedKeyExpr a True) <> ", got " <> show keyRes
+
+fetchKeyChangedExprSameValue : Test
+fetchKeyChangedExprSameValue = test $ do
+  a <- allocateParty "alice"
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a False)
+  foundContract <- a `submit` createAndExerciseCmd (V2.ChangedKeyExprHelper a) (V2.ChangedKeyExprFetch $ coerceContractId cid)
+  foundContract === V2.ChangedKeyExpr a False
+
+fetchKeyChangedExprChangedValue : Test
+fetchKeyChangedExprChangedValue = test $ do
+  a <- allocateParty "alice"
+  cid <- a `submit` createExactCmd (V1.ChangedKeyExpr a True)
+  {-
+  The fetch inside the following command correctly fails with:
```

Comment:
@samuel-williams-da @paulbrauner-da, it seems that plain `fetch` _does_ perform the upgrade validation

Reply:
I would assert that specific failure in the test. for the cases that are broken, write the correct logic and mark them in the test list